File: context-engineering-intro/frontend/src/lib/api/telnyx.ts
/**
 * Telnyx SMS/MMS API client
 *
 * Features:
 * - Telnyx settings management
 * - Phone number search and purchase
 * - SMS sending profiles
 * - 10DLC Brand and Campaign registration
 */

import { api } from '@/lib/api'

// =========================================================================
// TYPES
// =========================================================================

export interface TelnyxSettings {
  id: string
  object_id: string
  object_name?: string
  api_key_masked: string
  messaging_profile_id?: string
  webhook_secret_set: boolean
  is_active: boolean
  created_at: string
  updated_at: string
}

export interface CreateTelnyxSettingsRequest {
  object_id: string
  api_key: string
  messaging_profile_id?: string
  webhook_secret?: string
}

export interface UpdateTelnyxSettingsRequest {
  api_key?: string
  messaging_profile_id?: string
  webhook_secret?: string
  is_active?: boolean
}

export interface VerifyCredentialsResponse {
  valid: boolean
  balance?: string
  currency?: string
  error?: string
}

export interface AvailableNumber {
  phone_number: string
  type: string
  region?: string
  locality?: string
  features: string[]
  monthly_cost?: string
  upfront_cost?: string
}

export interface OwnedPhoneNumber {
  id: string
  phone_number: string
  friendly_name?: string
  external_id?: string
  messaging_profile_id?: string
  phone_number_type?: string
  capabilities: Record<string, any>
  campaign_id?: string
  campaign_name?: string
  status: string
  object_id: string
  object_name?: string
  purchased_at?: string
  created_at: string
}

export interface SearchNumbersRequest {
  object_id: string
  area_code?: string
  city?: string
  state?: string
  number_type?: 'local' | 'toll_free'
  features?: string[]
  limit?: number
}

export interface PurchaseNumberRequest {
  object_id: string
  phone_number: string
  friendly_name?: string
}

export interface SMSProfile {
  id: string
  object_id: string
  object_name?: string
  phone_number_id: string
  phone_number: string
  phone_number_friendly_name?: string
  display_name: string
  signature?: string
  max_messages_per_day?: string
  is_active: boolean
  created_at: string
  updated_at: string
  assigned_user_count: number
}

export interface AssignedSMSProfile {
  id: string
  phone_number: string
  display_name: string
  signature?: string
  is_default: boolean
}

export interface CreateSMSProfileRequest {
  object_id: string
  phone_number_id: string
  display_name: string
  signature?: string
  max_messages_per_day?: string
}

export interface Brand {
  id: string
  object_id: string
  object_name?: string
  external_brand_id?: string
  company_name: string
  display_name?: string
  ein?: string
  website?: string
  phone?: string
  email?: string
  street?: string
  city?: string
  state?: string
  postal_code?: string
  country: string
  vertical?: string
  entity_type?: string
  status: string
  vetting_status?: string
  vetting_score?: string
  rejection_reason?: string
  created_at: string
  updated_at: string
}

export interface CreateBrandRequest {
  object_id: string
  company_name: string
  display_name?: string
  ein?: string
  website?: string
  phone?: string
  email?: string
  street?: string
  city?: string
  state?: string
  postal_code?: string
  country?: string
  vertical?: string
  entity_type?: string
}

export interface Campaign {
  id: string
  brand_id: string
  brand_name?: string
  external_campaign_id?: string
  name: string
  description?: string
  use_case: string
  sub_use_case?: string
  sample_messages: string[]
  opt_in_message?: string
  opt_out_message?: string
  help_message?: string
  subscriber_optin: boolean
  subscriber_optout: boolean
  subscriber_help: boolean
  number_pool: boolean
  embedded_link: boolean
  embedded_phone: boolean
  status: string
  rejection_reason?: string
  phone_number_count: number
  created_at: string
  updated_at: string
}

export interface CreateCampaignRequest {
  brand_id: string
  name: string
  description?: string
  use_case: string
  sub_use_case?: string
  sample_messages: string[]
  opt_in_message?: string
  opt_out_message?: string
  help_message?: string
  subscriber_optin?: boolean
  subscriber_optout?: boolean
  subscriber_help?: boolean
  number_pool?: boolean
  embedded_link?: boolean
  embedded_phone?: boolean
}

// =========================================================================
// TELNYX SETTINGS API
// =========================================================================

export const telnyxSettingsApi = {
  /**
   * List Telnyx settings (Owner/Admin only)
   */
  list: async (objectId?: string): Promise<TelnyxSettings[]> => {
    const params = objectId ? { object_id: objectId } : {}
    const response = await api.get('/telnyx-settings/', { params })
    return response.data
  },

  /**
   * Get specific Telnyx settings
   */
  get: async (settingsId: string): Promise<TelnyxSettings> => {
    const response = await api.get(`/telnyx-settings/${settingsId}`)
    return response.data
  },

  /**
   * Create Telnyx settings for an object
   */
  create: async (data: CreateTelnyxSettingsRequest): Promise<TelnyxSettings> => {
    const response = await api.post('/telnyx-settings/', data)
    return response.data
  },

  /**
   * Update Telnyx settings
   */
  update: async (settingsId: string, data: UpdateTelnyxSettingsRequest): Promise<TelnyxSettings> => {
    const response = await api.put(`/telnyx-settings/${settingsId}`, data)
    return response.data
  },

  /**
   * Delete Telnyx settings
   */
  delete: async (settingsId: string): Promise<void> => {
    await api.delete(`/telnyx-settings/${settingsId}`)
  },

  /**
   * Verify Telnyx credentials
   */
  verify: async (settingsId: string): Promise<VerifyCredentialsResponse> => {
    const response = await api.post(`/telnyx-settings/${settingsId}/verify`)
    return response.data
  },
}

// =========================================================================
// PHONE NUMBER API
// =========================================================================

export const phoneNumbersApi = {
  /**
   * Search available phone numbers
   */
  search: async (data: SearchNumbersRequest): Promise<AvailableNumber[]> => {
    const response = await api.post('/phone-numbers/search', data)
    return response.data
  },

  /**
   * Purchase a phone number
   */
  purchase: async (data: PurchaseNumberRequest): Promise<OwnedPhoneNumber> => {
    const response = await api.post('/phone-numbers/purchase', data)
    return response.data
  },

  /**
   * List owned phone numbers
   */
  list: async (objectId?: string, status?: string): Promise<OwnedPhoneNumber[]> => {
    const params: Record<string, string> = {}
    if (objectId) params.object_id = objectId
    if (status) params.status_filter = status
    const response = await api.get('/phone-numbers/', { params })
    return response.data
  },

  /**
   * Get specific phone number
   */
  get: async (numberId: string): Promise<OwnedPhoneNumber> => {
    const response = await api.get(`/phone-numbers/${numberId}`)
    return response.data
  },

  /**
   * Update phone number
   */
  update: async (numberId: string, data: { friendly_name?: string; campaign_id?: string }): Promise<OwnedPhoneNumber> => {
    const response = await api.put(`/phone-numbers/${numberId}`, data)
    return response.data
  },

  /**
   * Release (cancel) phone number
   */
  release: async (numberId: string): Promise<void> => {
    await api.delete(`/phone-numbers/${numberId}`)
  },

  /**
   * Sync phone number from Telnyx
   */
  sync: async (numberId: string): Promise<{ status: string; telnyx_status?: string }> => {
    const response = await api.post(`/phone-numbers/${numberId}/sync`)
    return response.data
  },
}

// =========================================================================
// SMS PROFILES API
// =========================================================================

export const smsProfilesApi = {
  /**
   * Get current user's assigned SMS profiles
   */
  getMyProfiles: async (): Promise<AssignedSMSProfile[]> => {
    const response = await api.get('/sms-profiles/my-profiles')
    return response.data
  },

  /**
   * List all SMS profiles (Owner/Admin)
   */
  list: async (objectId?: string, isActive?: boolean): Promise<SMSProfile[]> => {
    const params: Record<string, any> = {}
    if (objectId) params.object_id = objectId
    if (isActive !== undefined) params.is_active = isActive
    const response = await api.get('/sms-profiles/', { params })
    return response.data
  },

  /**
   * Get specific SMS profile with assigned users
   */
  get: async (profileId: string): Promise<{ profile: SMSProfile; assigned_users: any[] }> => {
    const response = await api.get(`/sms-profiles/${profileId}`)
    return response.data
  },

  /**
   * Create SMS profile
   */
  create: async (data: CreateSMSProfileRequest): Promise<SMSProfile> => {
    const response = await api.post('/sms-profiles/', data)
    return response.data
  },

  /**
   * Update SMS profile
   */
  update: async (profileId: string, data: Partial<CreateSMSProfileRequest> & { is_active?: boolean }): Promise<SMSProfile> => {
    const response = await api.put(`/sms-profiles/${profileId}`, data)
    return response.data
  },

  /**
   * Delete SMS profile
   */
  delete: async (profileId: string): Promise<void> => {
    await api.delete(`/sms-profiles/${profileId}`)
  },

  /**
   * Assign users to profile
   */
  assignUsers: async (profileId: string, assignments: { user_id: string; is_default?: boolean }[]): Promise<void> => {
    await api.post(`/sms-profiles/${profileId}/assignments`, assignments)
  },

  /**
   * Remove user assignment
   */
  removeAssignment: async (profileId: string, userId: string): Promise<void> => {
    await api.delete(`/sms-profiles/${profileId}/assignments/${userId}`)
  },

  /**
   * Set default profile for user
   */
  setDefault: async (profileId: string, userId: string): Promise<void> => {
    await api.put(`/sms-profiles/${profileId}/assignments/${userId}/default`)
  },
}

// =========================================================================
// 10DLC API
// =========================================================================

export const tenDLCApi = {
  /**
   * List 10DLC brands
   */
  listBrands: async (objectId?: string): Promise<Brand[]> => {
    const params = objectId ? { object_id: objectId } : {}
    const response = await api.get('/10dlc/brands', { params })
    return response.data
  },

  /**
   * Create 10DLC brand
   */
  createBrand: async (data: CreateBrandRequest): Promise<Brand> => {
    const response = await api.post('/10dlc/brands', data)
    return response.data
  },

  /**
   * Sync brand status from Telnyx
   */
  syncBrand: async (brandId: string): Promise<{ status: string; brand_status?: string }> => {
    const response = await api.post(`/10dlc/brands/${brandId}/sync`)
    return response.data
  },

  /**
   * List 10DLC campaigns
   */
  listCampaigns: async (brandId?: string): Promise<Campaign[]> => {
    const params = brandId ? { brand_id: brandId } : {}
    const response = await api.get('/10dlc/campaigns', { params })
    return response.data
  },

  /**
   * Create 10DLC campaign
   */
  createCampaign: async (data: CreateCampaignRequest): Promise<Campaign> => {
    const response = await api.post('/10dlc/campaigns', data)
    return response.data
  },

  /**
   * Assign phone number to campaign
   */
  assignNumber: async (campaignId: string, phoneNumberId: string): Promise<{ status: string }> => {
    const response = await api.post(`/10dlc/campaigns/${campaignId}/assign-number?phone_number_id=${phoneNumberId}`)
    return response.data
  },

  /**
   * Sync campaign status from Telnyx
   */
  syncCampaign: async (campaignId: string): Promise<{ status: string; campaign_status?: string }> => {
    const response = await api.post(`/10dlc/campaigns/${campaignId}/sync`)
    return response.data
  },
}

File: context-engineering-intro/backend/app/models/telnyx.py
"""
Telnyx SMS/MMS integration models

Models for:
- Tenant-level Telnyx API settings
- Phone number inventory
- 10DLC Brand registration
- 10DLC Campaign registration
"""

from sqlalchemy import Column, String, DateTime, Boolean, ForeignKey, Index, Text, Enum as SQLEnum
from sqlalchemy.dialects.postgresql import UUID, JSONB
from sqlalchemy.orm import relationship
from sqlalchemy.sql import func
import uuid
import enum

from app.config.database import Base


class TelnyxBrandStatus(str, enum.Enum):
    """Status of 10DLC brand registration"""
    DRAFT = "draft"
    PENDING = "pending"
    VERIFIED = "verified"
    REJECTED = "rejected"
    FAILED = "failed"


class TelnyxCampaignStatus(str, enum.Enum):
    """Status of 10DLC campaign registration"""
    DRAFT = "draft"
    PENDING = "pending"
    APPROVED = "approved"
    REJECTED = "rejected"
    SUSPENDED = "suspended"


class TelnyxPhoneNumberStatus(str, enum.Enum):
    """Status of phone number"""
    ACTIVE = "active"
    PENDING = "pending"
    RELEASED = "released"
    FAILED = "failed"


class ObjectTelnyxSettings(Base):
    """
    Telnyx API settings per Object (tenant).

    Stores encrypted API keys and configuration for Telnyx integration.
    Each Object can have its own Telnyx account.
    """
    __tablename__ = "object_telnyx_settings"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)

    # Link to Object (tenant)
    object_id = Column(UUID(as_uuid=True), ForeignKey("objects.id", ondelete="CASCADE"),
                      nullable=False, index=True)

    # Telnyx API credentials (encrypted)
    api_key = Column(Text, nullable=False)  # Encrypted using encryption_key

    # Messaging Profile ID (required for sending)
    messaging_profile_id = Column(String(255), nullable=True)

    # Webhook configuration
    webhook_secret = Column(String(255), nullable=True)  # For signature verification

    # Status
    is_active = Column(Boolean, nullable=False, default=True)

    # Timestamps
    created_at = Column(DateTime(timezone=True), server_default=func.now(), nullable=False)
    updated_at = Column(DateTime(timezone=True), server_default=func.now(), onupdate=func.now(),
                       nullable=False)

    # Relationships
    object = relationship("Object", backref="telnyx_settings")
    phone_numbers = relationship("TelnyxPhoneNumber", back_populates="telnyx_settings",
                                cascade="all, delete-orphan")
    brands = relationship("TelnyxBrand", back_populates="telnyx_settings",
                         cascade="all, delete-orphan")

    # Indexes
    __table_args__ = (
        Index('idx_telnyx_settings_object', 'object_id'),
        Index('idx_telnyx_settings_active', 'is_active'),
    )

    def __repr__(self):
        return f"<ObjectTelnyxSettings object_id={self.object_id}>"


class TelnyxPhoneNumber(Base):
    """
    Phone numbers owned via Telnyx.

    Tracks phone numbers purchased through Telnyx API for SMS/MMS.
    """
    __tablename__ = "telnyx_phone_numbers"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)

    # Link to Object (tenant)
    object_id = Column(UUID(as_uuid=True), ForeignKey("objects.id", ondelete="CASCADE"),
                      nullable=False, index=True)

    # Link to Telnyx settings
    telnyx_settings_id = Column(UUID(as_uuid=True),
                                ForeignKey("object_telnyx_settings.id", ondelete="CASCADE"),
                                nullable=False, index=True)

    # Phone number details
    phone_number = Column(String(20), nullable=False, unique=True)  # E.164 format: +1234567890
    external_id = Column(String(255), nullable=True, index=True)  # Telnyx phone number ID

    # Telnyx-specific fields
    messaging_profile_id = Column(String(255), nullable=True)  # Assigned messaging profile
    phone_number_type = Column(String(50), nullable=True)  # local, toll-free, etc.

    # Capabilities (what this number can do)
    capabilities = Column(JSONB, default=dict)
    # Example: {"sms": {"domestic_two_way": true}, "mms": {"domestic_two_way": true}, "voice": true}

    # 10DLC campaign assignment
    campaign_id = Column(UUID(as_uuid=True), ForeignKey("telnyx_campaigns.id", ondelete="SET NULL"),
                        nullable=True, index=True)

    # Display name for UI
    friendly_name = Column(String(255), nullable=True)

    # Status
    status = Column(SQLEnum(TelnyxPhoneNumberStatus), default=TelnyxPhoneNumberStatus.PENDING)

    # Timestamps
    purchased_at = Column(DateTime(timezone=True), nullable=True)
    released_at = Column(DateTime(timezone=True), nullable=True)
    created_at = Column(DateTime(timezone=True), server_default=func.now(), nullable=False)
    updated_at = Column(DateTime(timezone=True), server_default=func.now(), onupdate=func.now(),
                       nullable=False)

    # Relationships
    object = relationship("Object", backref="telnyx_phone_numbers")
    telnyx_settings = relationship("ObjectTelnyxSettings", back_populates="phone_numbers")
    campaign = relationship("TelnyxCampaign", back_populates="phone_numbers")
    sms_profiles = relationship("SMSSendingProfile", back_populates="phone_number",
                               cascade="all, delete-orphan")

    # Indexes
    __table_args__ = (
        Index('idx_phone_number_object', 'object_id'),
        Index('idx_phone_number_status', 'status'),
        Index('idx_phone_number_campaign', 'campaign_id'),
    )

    def __repr__(self):
        return f"<TelnyxPhoneNumber {self.phone_number}>"


class TelnyxBrand(Base):
    """
    10DLC Brand registration for A2P SMS compliance.

    Required for sending SMS in the US. Brands must be vetted
    before campaigns can be created.
    """
    __tablename__ = "telnyx_brands"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)

    # Link to Object (tenant)
    object_id = Column(UUID(as_uuid=True), ForeignKey("objects.id", ondelete="CASCADE"),
                      nullable=False, index=True)

    # Link to Telnyx settings
    telnyx_settings_id = Column(UUID(as_uuid=True),
                                ForeignKey("object_telnyx_settings.id", ondelete="CASCADE"),
                                nullable=False, index=True)

    # Telnyx external ID
    external_brand_id = Column(String(255), nullable=True, index=True)  # From Telnyx API

    # Company Information (required for registration)
    company_name = Column(String(255), nullable=False)
    display_name = Column(String(255), nullable=True)  # Shown to recipients
    ein = Column(String(20), nullable=True)  # Employer Identification Number

    # Contact Information
    website = Column(String(500), nullable=True)
    phone = Column(String(20), nullable=True)
    email = Column(String(255), nullable=True)

    # Address
    street = Column(String(255), nullable=True)
    city = Column(String(100), nullable=True)
    state = Column(String(50), nullable=True)
    postal_code = Column(String(20), nullable=True)
    country = Column(String(2), nullable=False, default="US")  # ISO country code

    # Business Details
    vertical = Column(String(100), nullable=True)  # Industry vertical
    entity_type = Column(String(50), nullable=True)  # PRIVATE, PUBLIC, NON_PROFIT, GOVERNMENT
    stock_symbol = Column(String(10), nullable=True)  # If public company
    stock_exchange = Column(String(50), nullable=True)

    # Registration Status
    status = Column(SQLEnum(TelnyxBrandStatus), default=TelnyxBrandStatus.DRAFT)

    # Vetting Results
    vetting_status = Column(String(50), nullable=True)
    vetting_score = Column(String(10), nullable=True)  # e.g., "A", "B", "C"
    vetting_date = Column(DateTime(timezone=True), nullable=True)
    rejection_reason = Column(Text, nullable=True)

    # Raw response from Telnyx (for debugging)
    provider_metadata = Column(JSONB, default=dict)

    # Timestamps
    created_at = Column(DateTime(timezone=True), server_default=func.now(), nullable=False)
    updated_at = Column(DateTime(timezone=True), server_default=func.now(), onupdate=func.now(),
                       nullable=False)

    # Relationships
    object = relationship("Object", backref="telnyx_brands")
    telnyx_settings = relationship("ObjectTelnyxSettings", back_populates="brands")
    campaigns = relationship("TelnyxCampaign", back_populates="brand",
                           cascade="all, delete-orphan")

    # Indexes
    __table_args__ = (
        Index('idx_brand_object', 'object_id'),
        Index('idx_brand_status', 'status'),
        Index('idx_brand_external', 'external_brand_id'),
    )

    def __repr__(self):
        return f"<TelnyxBrand {self.company_name}>"


class TelnyxCampaign(Base):
    """
    10DLC Campaign registration for A2P SMS compliance.

    Campaigns are tied to brands and define the use case for messaging.
    Phone numbers must be assigned to approved campaigns.
    """
    __tablename__ = "telnyx_campaigns"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)

    # Link to Brand
    brand_id = Column(UUID(as_uuid=True), ForeignKey("telnyx_brands.id", ondelete="CASCADE"),
                     nullable=False, index=True)

    # Telnyx external ID
    external_campaign_id = Column(String(255), nullable=True, index=True)  # From Telnyx API

    # Campaign Details
    name = Column(String(255), nullable=False)
    description = Column(Text, nullable=True)

    # Use Case (required)
    use_case = Column(String(100), nullable=False)
    # Options: ACCOUNT_NOTIFICATION, CUSTOMER_CARE, DELIVERY_NOTIFICATION,
    # FRAUD_ALERT, HIGHER_EDUCATION, MARKETING, MIXED, POLLING_VOTING,
    # PUBLIC_SERVICE_ANNOUNCEMENT, SECURITY_ALERT, 2FA

    # Sub Use Case (optional, more specific)
    sub_use_case = Column(String(100), nullable=True)

    # Sample Messages (required for review)
    sample_messages = Column(JSONB, default=list)
    # Example: ["Your order has shipped!", "Your verification code is {code}"]

    # Keywords configuration
    opt_in_keywords = Column(JSONB, default=list)  # e.g., ["START", "YES"]
    opt_out_keywords = Column(JSONB, default=list)  # e.g., ["STOP", "UNSUBSCRIBE"]
    help_keywords = Column(JSONB, default=list)  # e.g., ["HELP", "INFO"]

    # Message Templates
    opt_in_message = Column(Text, nullable=True)  # Sent when user opts in
    opt_out_message = Column(Text, nullable=True)  # Sent when user opts out
    help_message = Column(Text, nullable=True)  # Sent when user texts HELP

    # Campaign Attributes
    subscriber_optin = Column(Boolean, default=True)  # Opt-in required?
    subscriber_optout = Column(Boolean, default=True)  # Opt-out available?
    subscriber_help = Column(Boolean, default=True)  # Help available?
    number_pool = Column(Boolean, default=False)  # Multiple numbers in campaign?
    direct_lending = Column(Boolean, default=False)  # Direct lending messages?
    embedded_link = Column(Boolean, default=False)  # Links in messages?
    embedded_phone = Column(Boolean, default=False)  # Phone numbers in messages?
    affiliate_marketing = Column(Boolean, default=False)  # Affiliate marketing?
    age_gated = Column(Boolean, default=False)  # Age-restricted content?

    # Status
    status = Column(SQLEnum(TelnyxCampaignStatus), default=TelnyxCampaignStatus.DRAFT)
    rejection_reason = Column(Text, nullable=True)

    # Raw response from Telnyx (for debugging)
    provider_metadata = Column(JSONB, default=dict)

    # Timestamps
    submitted_at = Column(DateTime(timezone=True), nullable=True)
    approved_at = Column(DateTime(timezone=True), nullable=True)
    created_at = Column(DateTime(timezone=True), server_default=func.now(), nullable=False)
    updated_at = Column(DateTime(timezone=True), server_default=func.now(), onupdate=func.now(),
                       nullable=False)

    # Relationships
    brand = relationship("TelnyxBrand", back_populates="campaigns")
    phone_numbers = relationship("TelnyxPhoneNumber", back_populates="campaign")

    # Indexes
    __table_args__ = (
        Index('idx_campaign_brand', 'brand_id'),
        Index('idx_campaign_status', 'status'),
        Index('idx_campaign_external', 'external_campaign_id'),
    )

    def __repr__(self):
        return f"<TelnyxCampaign {self.name}>"
